FrEee/Scripts/AI/Vanguard/MinistryOfInfrastructure.csx

```diff
@@ -0,0 +1,246 @@
+/// </summary>
+public class MinistryOfInfrastructure
+{
+    public Empire Empire;
+
+    public Galaxy Galaxy;
+
+    public List<FacilityTemplate> AvailableFacilities; 
+
+
+    public void Run(Empire empire, Galaxy galaxy)
+    {
+        Empire = empire;
+        Galaxy = galaxy;
+        //get the latest versions of each facility. 
+        AvailableFacilities = Empire.Current.UnlockedItems.OfType<FacilityTemplate>()
+            .Where(f => f.ID == 0 && f.Cost.Any()).OnlyLatestVersions(f => f.Family).ToList();
+
+        ConstructNewFacilities();
+
+    }
+
+
+    /// <summary>
+    /// Construct facilities on any world where there are empty slots. 
+    /// </summary>
+    protected void ConstructNewFacilities()
+    {
+        var emptySlots = Empire.ColonizedPlanets.Where(x => x.Colony.ConstructionQueue.FacilitySlotsFree > 0); 
+
+        foreach(var planet in emptySlots)
+        {
+            var colony = planet.Colony; 
+            while(colony.ConstructionQueue.FacilitySlotsFree > 0)
+            {
+                var chosenFacility = GetEmpireVitalFaciltiesForSystem(planet.StarSystem);
+                if (chosenFacility == null)
+                    chosenFacility = GetPlanetVitalFacilities(planet);
+                if (chosenFacility == null)
+                    chosenFacility = GetResourceFacility(planet);
+                if (chosenFacility == null)
+                    continue; 
+                
+
+                var order = new ConstructionOrder<Facility, FacilityTemplate> { Template = chosenFacility };
+                colony.ConstructionQueue.Orders.Add(order);
+                var cmd = new AddOrderCommand
+                (
+                    colony.ConstructionQueue,
+                    order
+                );
+                Empire.Commands.Add(cmd);
+                Empire.LogAIMessage($"Construction of a {chosenFacility.Name} on {planet.Name} has been ordered");
+            }
+        }
+    }
+
+
+    /// <summary>
+    /// Ensure that all vital system options are taken care of. 
+    /// </summary>
+    /// <param name="starSystem"></param>
+    /// <returns></returns>
+    protected FacilityTemplate GetEmpireVitalFaciltiesForSystem(StarSystem starSystem)
+    {
+        //spaceports. 
+        var option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Spaceport"));
+        if (option != null && !HasStarSystemFacility(starSystem, "Spaceport"))
+            return option;
+        //resupply. 
+        option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Supply Generation"));
+        if (option != null && !HasStarSystemFacility(starSystem, "Supply Generation"))
+            return option;
+        //spaceyard. Make sure every system has at least one. 
+        //TODO; mod might make more than one facility be a space yard... make sure to select the right one. 
+        option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Space Yard"));
```

Comment:
Only one space yard per star system?

Reply:
At least one space yard. Will later figure out if more are needed when I build some proper strategy AI. Plus, orbital space yards will be a thing.

Reply:
OK, I guess that's good enough for now then! Typically I build a space yard on every colony but maybe I'm in the minority... :P